### PR TITLE
feat: exclude hidden subfolders by default in path-input autocomplete

### LINKED ISSUE #321: Hide dot files from the autocomplete in the address bar
### Is your request due to an existing bug?

Currently, the autocomplete of the address bar show hidden files by default.

Image

### What feature do you want in rovr?

 It should be hidden and only shown when I start to type "."

### Describe any alternatives you've considered

_No response_

### Are there any other file managers with this feature?

Nautilus (GNOME File Manager) does it.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hongquan**
But include them when users really want to find them, by typing the prefix
of their names.

Resolves #321 

Normal state: Exclude ".git"

image

When user wants to find hidden subfolders:

image

---

by submitting this pull request, i agree that

- [x] i have run `poe check` to check for any style issues and fixed them
- [x] i have tested rovr (and also ran `poe test` if applicable) to make sure my changes do not break anything
- [x] cache, logs, dotfiles and/or others were not accidentally added to git's tracking history
- [x] my commits (or at least the pr title) follow the conventional commits format as much as possible

## Summary by Sourcery

Adjust path autocomplete to intelligently include or exclude hidden subdirectories based on the user’s typed prefix across Unix and Windows.

Enhancements:
- Introduce cross-platform hidden-directory detection and filtering in path autocomplete suggestions, allowing explicit discovery via dot-prefixed input.
- Refine Unix and Windows candidate generation to use scandir/DirEntry, handle navigation tokens like "..", and improve exact-directory matching behavior.

Tests:
- Add unit tests for the hidden-filtering decision logic to cover dotfile prefixes, navigation components, separators, and trailing-slash edge cases.

## Summary by Sourcery

Adjust path autocomplete to filter hidden subdirectories by default while still allowing users to discover them via dot-prefixed input, with consistent behavior across Unix and Windows.

New Features:
- Enable path-input autocomplete to reveal hidden subdirectories when the user types a dot-prefixed component in the path.

Enhancements:
- Introduce cross-platform hidden-directory detection using DirEntry metadata and dot-prefix rules in autocomplete suggestions.
- Refine Unix and Windows autocomplete candidate generation to use scandir/Path, handle exact-directory matches, and ignore suggestions while navigating with "..".

Tests:
- Add unit tests covering the hidden-filtering decision logic, including dotfile prefixes, navigation tokens, separators, and trailing-slash edge cases.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved path autocomplete handling for missing or inaccessible directories.
  - Hidden directories are now excluded by default, while remaining available when explicitly requested.
  - Path navigation and parent-directory completion now behave consistently across Windows and Unix systems.
  - Suppressed unnecessary `..` suggestions.

- **Tests**
  - Added comprehensive coverage for hidden paths, separators, parent traversal, and trailing slashes across supported platforms.

### CODE REVIEW

File: src/rovr/navigation_widgets/path_input.py [hunk 1]
```diff
@@ -11,23 +14,82 @@
...
+        ``True`` when hidden entries should be filtered, ``False`` when they
+        should be listed.
+    """
+    # Never filter when the raw input is empty
+    if not path_str:
+        return False
+
+    # Navigation tokens never reveal hidden files.
+    if path_str.endswith(".."):
+        return True
```
[the remark this answers is not in the document]
Comment (hongquan) [reply src/rovr/navigation_widgets/path_input.py:79]: No need, if the input is "/home/quan/Works/../rovr", it fallback to the last check with `os.path.split` and still work as expected.

File: src/rovr/navigation_widgets/path_input.py [hunk 2]
```diff
@@ -82,28 +170,31 @@ def _win_get_candidates(path_str: str) -> list[DropdownItem]:
...
-            if path.isdir(full_item_path):
-                items.append(PathDropdownItem(item + "/", full_item_path))
-        items.sort(key=lambda x: x.path.lower())
-        return items
-    return []
+        return [PathDropdownItem(f"{ppath.name}{os.path.sep}", str(ppath))]
+
+    # Case 4: list contents of parent
+    # - Path ends with "/" → show subdirectories
+    # - Partial path like "C:\Users\Admin\Doc" → parent is "C:\Users\Admin"
```
Comment (NSPC911) [inline src/rovr/navigation_widgets/path_input.py]:
```suggestion
    # - Path ends with "/" -> show subdirectories
    # - Partial path like "C:\Users\Admin\Doc" -> parent is "C:\Users\Admin"
```
really stupid nitpick

File: src/rovr/navigation_widgets/path_input.py [hunk 3]
```diff
@@ -11,23 +14,81 @@
...
     except OSError:
-        return []
+        return False
+
+    if os_type == "Darwin":
+        return file_stat.st_flags & stat.UF_HIDDEN
+    # Windows
+    if os_type == "Windows":
+        return file_stat.st_file_attributes & stat.FILE_ATTRIBUTE_HIDDEN
+    return False
```
Comment (NSPC911) [inline src/rovr/navigation_widgets/path_input.py:52]: A function already exists in the `rovr.functions.path` module; it is an unnecessary extra function.

Comment (hongquan) [reply src/rovr/navigation_widgets/path_input.py:52]: Hi, I saw the function in `rovr.functions.path` but not use it because:
- I want to treat ".git" as hidden on Windows also.
- The function `rovr.functions.path` is not well implemented, like: 
 
Why it needs:

```py
if os_type == "Windows":
    _GetFileAttributesW = ctypes.windll.kernel32.GetFileAttributesW
    _GetFileAttributesW.argtypes = [ctypes.c_wchar_p]
    _GetFileAttributesW.restype = ctypes.c_uint32
attrs = _GetFileAttributesW(filepath)
if attrs == 0xFFFFFFFF:  # INVALID_FILE_ATTRIBUTES
return bool(attrs & 0x02)  # FILE_ATTRIBUTE_HIDDEN
```

Why it needs:
```py
getattr(filepath.stat(follow_symlinks=False), "st_flags", 0) & getattr(stat, "UF_HIDDEN", 0)
```

**Comment by NSPC911** [comment]
I like the overall idea, but here are two concerns I have.
1. Hidden folders (like dotfolders and other versions on other OSes) are invisible no matter the user's preference. To make it more explicit, I would make it depend on `config["interface"]["show_hidden_files"]` on whether to show such hidden folders
2. Excluding hidden folders on other OSes is bad behaviour. For example, you know that there are dotfolder at a given directory, so you can do `.` in the input to view hidden folders instead (in this PR). However, the same cannot be done on other platforms, because you simply do not know the names of the hidden files. There doesn't seem to be any file explorers that do something similar to what is suggested in this PR

**Comment by NSPC911** [comment]
I understand the vision behind this, it's just that I think it would be fitting to keep it a linux behaviour than a cross platform behaviour.

**Comment by hongquan** [comment]
> I would make it depend on config["interface"]["show_hidden_files"]

For the path-input autocomplete, I do not rely on `config["interface"]["show_hidden_files"]`, because:
- I want the autocomplete list not inflate. You know, in this AI age, many new apps create a lot of dot folders (they create _~/.xxx_ instead of saving config to _~/.config/xxx_, per Linux standard).
- In practice, when using Nautilus (GNOME File Manager), I try to keep my working folder tidy by:
   + Let Nautilus hide hidden items most of the time.
   + Whenever I want to visit a hidden folder (_.git_, _.helix_), I do it from the address bar.

By that, I don't have to always see the unrelated, busy dot items.

**Comment by hongquan** [comment]
> it's just that I think it would be fitting to keep it a linux behaviour than a cross platform behaviour

I am willing to revert the behavior of treating ".git" as hidden on Windows. Just need to hear your opinion, a Windows user, about it.

**Comment by NSPC911** [comment]
>     * I want the autocomplete list not inflate. You know, in this AI age, many new apps create a lot of dot folders (they create `~/.xxx` instead of saving config to `~/.config/xxx`, per Linux standard).
> 
>     * In practice, when using Nautilus (GNOME File Manager), I try to keep my working folder tidy by:
>       
>       * Let Nautilus hide hidden items most of the time.
>       * Whenever I want to visit a hidden folder (`.git`, `.helix`), I do it from the address bar.
> 
> By that, I don't have to always see the unrelated, busy dot items.

I was suggesting a mix.<br>
Say `show_hidden_files` is disabled, you will only see non-dotfolders. When you type in the input `/home/<user>/.`, you will then see the dotfolders.<br>
When you have `show_hidden_files` enabled, you will see dotfolders regardless.

This does something similar to what you stated: Hide hidden items, visit only via the path input.

**Comment by hongquan** [comment]
> When you have show_hidden_files enabled, you will see dotfolders regardless.

Ah, yes, I will do that.

**Review by NSPC911 — APPROVED** [review]
approved

**Comment by hongquan** [comment]
How to use that "Approved" sticker? I want to apply in my projects.

**Comment by NSPC911** [comment]
Haha, it is
```md
![approved](https://media1.tenor.com/m/lztbLMEBDs4AAAAC/sonic-boom-knuckles.gif)
```
